src/Quaternion.jl

```diff
@@ -24,10 +24,33 @@ promote_rule(::Type{Quaternion{T}}, ::Type{S}) where {T <: Real, S <: Real} = Qu
 promote_rule(::Type{Quaternion{T}}, ::Type{Complex{S}}) where {T <: Real, S <: Real} = Quaternion{promote_type(T, S)}
 promote_rule(::Type{Quaternion{T}}, ::Type{Quaternion{S}}) where {T <: Real, S <: Real} = Quaternion{promote_type(T, S)}
 
+"""
+    quat(r, [i, j, k])
+
+Convert real numbers or arrays to quaternion. `i, j, k` defaults to zero.
+
+# Examples
+```jldoctest
+julia> quat(7)
+Quaternion{Int64}(7, 0, 0, 0, false)
+
+julia> quat([1, 2, 3])
+3-element Vector{Quaternion{Int64}}:
+  Quaternion{Int64}(1, 0, 0, 0, true)
+ Quaternion{Int64}(2, 0, 0, 0, false)
+ Quaternion{Int64}(3, 0, 0, 0, false)
+```
+"""
 quat(p, v1, v2, v3) = Quaternion(p, v1, v2, v3)
 quat(p, v1, v2, v3, n) = Quaternion(p, v1, v2, v3, n)
 quat(x) = Quaternion(x)
 quat(s, a) = Quaternion(s, a)
+function quat(A::AbstractArray{T}) where T
+    if !isconcretetype(T)
+        error("`quat` not defined on abstractly-typed arrays; please convert to a more specific type")
+    end
+    convert(AbstractArray{typeof(quat(zero(T)))}, A)
+end
```

Comment:
This implementation is the same as 
https://github.com/JuliaLang/julia/blob/1ece3f64e093ba47bcd5e7a01193b8c9b05d441e/base/complex.jl#L1114-L1119